File: src/main/java/com/king/java/field/plugin/dialog/GenerateFieldDialog.java
/*
 * Copyright (C) 2020 Jenly Yu, https://github.com/jenly1314/JavaFieldGenerator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.king.java.field.plugin.dialog;

import com.intellij.openapi.ui.Messages;
import com.king.java.field.plugin.component.ConfigComponent;
import com.king.java.field.plugin.entity.FieldParseConfig;
import org.apache.commons.lang.StringUtils;

import javax.swing.*;
import java.util.Arrays;

import static com.king.java.field.plugin.dialog.GenerateFieldDialog.Modifier.*;

/**
 * 生成字段对话框
 *
 * @author <a href="mailto:yujinlin@mail.com">Jenly</a>
 */
public class GenerateFieldDialog extends JDialog {

    private JPanel panel;
    private JTextArea textArea;
    private JButton btnCancel, btnGenerate, btnSetting;
    private OnClickListener onClickListener;
    private JRadioButton rbPublic, rbProtected, rbPrivate;
    private JCheckBox cbGetAndSet;
    private JCheckBox cbToString;
    private JSpinner spField;
    private JSpinner spFieldType;
    private JSpinner spFieldComment;

    private FieldParseConfig fieldParseConfig;

    private static final int MINIMUM = 0;
    private static final int MAXIMUM = 9;

    public enum Modifier {
        /**
         * 公开
         */
        PUBLIC("public"),
        /**
         * 受保护
         */
        PROTECTED("protected"),
        /**
         * 私有
         */
        PRIVATE("private");

        private String value;

        Modifier(String value) {
            this.value = value;
        }

        public String getValue() {
            return value;
        }

        public static Modifier ofValue(String value) {
            return Arrays.stream(values()).filter(it -> it.value.equals(value)).findFirst().get();
        }

    }

    /**
     * 成员变量可见性修饰符
     */
    private Modifier modifier;

    public GenerateFieldDialog() {
        setContentPane(panel);
        setModal(true);
        setTitle("Java Field");

        getRootPane().setDefaultButton(btnGenerate);

        fieldParseConfig = ConfigComponent.getInstance().getFieldParseConfig();
        modifier = ofValue(fieldParseConfig.getFieldModifier());
        switch (modifier) {
            case PUBLIC:
                rbPublic.setSelected(true);
                break;
            case PROTECTED:
                rbProtected.setSelected(true);
                break;
            case PRIVATE:
                rbPrivate.setSelected(true);
                break;
            default:
                break;
        }

        cbGetAndSet.setSelected(fieldParseConfig.isGenerateGetterAndSetter());
        cbToString.setSelected(fieldParseConfig.isGenerateToString());


        spField.addChangeListener(l -> {
            fieldParseConfig.setFieldColumn((int) spField.getValue());
        });

        spFieldType.addChangeListener(l -> {
            fieldParseConfig.setFieldTypeColumn((int) spFieldType.getValue());
        });

        spFieldComment.addChangeListener(l -> {
            fieldParseConfig.setFieldCommentColumn((int) spFieldComment.getValue());
        });

        rbPublic.addChangeListener(l -> {
            if (rbPublic.isSelected()) {
                modifier = PUBLIC;
                rbProtected.setSelected(false);
                rbPrivate.setSelected(false);
                fieldParseConfig.setFieldModifier(modifier.getValue());
            }
        });
        rbProtected.addChangeListener(l -> {
            if (rbProtected.isSelected()) {
                modifier = PROTECTED;
                rbPublic.setSelected(false);
                rbPrivate.setSelected(false);
                fieldParseConfig.setFieldModifier(modifier.getValue());
            }
        });
        rbPrivate.addChangeListener(l -> {
            if (rbPrivate.isSelected()) {
                modifier = PRIVATE;
                rbPublic.setSelected(false);
                rbProtected.setSelected(false);
                fieldParseConfig.setFieldModifier(modifier.getValue());
            }
        });

        cbGetAndSet.addChangeListener(l -> {
            fieldParseConfig.setGenerateGetterAndSetter(cbGetAndSet.isSelected());
        });
        cbToString.addChangeListener(l -> {
            fieldParseConfig.setGenerateToString(cbToString.isSelected());
        });

        btnGenerate.addActionListener(l -> {
            if (StringUtils.isNotBlank(textArea.getText())) {
                if (onClickListener != null) {
                    onClickListener.onGenerate(fieldParseConfig, textArea.getText());
                }
                dispose();
            } else {
                Messages.showMessageDialog("Text can not be null!", "Error", Messages.getInformationIcon());
            }

        });
        btnCancel.addActionListener(l -> {
            if (onClickListener != null) {
                onClickListener.onCancel();
            }
            dispose();
        });
        btnSetting.addActionListener(l -> {
            SettingFieldTypeDialog settingFieldTypeDialog = new SettingFieldTypeDialog();
            // 自动调整对话框大小
            settingFieldTypeDialog.pack();
            settingFieldTypeDialog.setLocationRelativeTo(this);
            settingFieldTypeDialog.setVisible(true);
        });

        spField.setModel(createSpinnerNumberModel());
        spFieldType.setModel(createSpinnerNumberModel());
        spFieldComment.setModel(createSpinnerNumberModel());

        spField.setValue(fieldParseConfig.getFieldColumn());
        spFieldType.setValue(fieldParseConfig.getFieldTypeColumn());
        spFieldComment.setValue(fieldParseConfig.getFieldCommentColumn());
    }

    private SpinnerNumberModel createSpinnerNumberModel(){
        SpinnerNumberModel spinnerNumberModel = new SpinnerNumberModel();
        spinnerNumberModel.setMinimum(MINIMUM);
        spinnerNumberModel.setMaximum(MAXIMUM);
        return spinnerNumberModel;
    }

    /**
     * 点击监听器
     */
    public interface OnClickListener {
        /**
         * 生成
         *
         * @param fieldParseConfig
         * @param text
         */
        void onGenerate(FieldParseConfig fieldParseConfig, String text);

        /**
         * 取消
         */
        void onCancel();
    }

    /**
     * 设置监听器
     *
     * @param onClickListener
     */
    public void setOnClickListener(OnClickListener onClickListener) {
        this.onClickListener = onClickListener;
    }
}

File: src/main/java/com/king/jvm/field/generator/ui/GenerateFieldDialog.kt
/*
 * Copyright (C) 2020 Jenly Yu, https://github.com/jenly1314/JvmFieldGenerator
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.king.jvm.field.generator.ui

import com.intellij.ide.BrowserUtil
import com.intellij.openapi.diagnostic.Logger
import com.intellij.openapi.ui.DialogWrapper
import com.intellij.openapi.ui.Messages
import com.intellij.util.ui.JBUI
import com.king.jvm.field.generator.component.ConfigComponent
import com.king.jvm.field.generator.model.FieldParseConfig
import java.awt.BorderLayout
import java.awt.Cursor
import java.awt.Dimension
import java.awt.event.MouseAdapter
import java.awt.event.MouseEvent
import java.util.regex.Pattern
import javax.swing.Box
import javax.swing.BoxLayout
import javax.swing.JButton
import javax.swing.JCheckBox
import javax.swing.JComponent
import javax.swing.JLabel
import javax.swing.JPanel
import javax.swing.JScrollPane
import javax.swing.JTextArea
import javax.swing.JTextField
import javax.swing.SwingConstants
import javax.swing.event.DocumentEvent
import javax.swing.event.DocumentListener

/**
 * 用于生成字段的主对话框
 *
 * @author <a href="mailto:yujinlin@mail.com">Jenly</a>
 * <p>
 * <a href="https://github.com/jenly1314">Follow me</a>
 */
class GenerateFieldDialog(initialClassName: String?) : DialogWrapper(true) {

    private val panel: JPanel
    private lateinit var jTextAreaInput: JTextArea
    private lateinit var tfClassName: JTextField
    private lateinit var chkRegexValidation: JCheckBox
    private lateinit var btnCancel: JButton
    private lateinit var btnGenerate: JButton
    private lateinit var btnSetting: JButton
    private var onClickListener: OnClickListener? = null

    private var fieldParseConfig: FieldParseConfig = ConfigComponent.getInstance().resolveFieldParseConfig()

    init {
        panel = initUiComponents(initialClassName)
        title = "JvmFieldGenerator - Generate Fields"
        init()
        registerEscapeAction()
        setupActions()
        applyPrimaryButtonStyle()
    }

    override fun createCenterPanel(): JComponent = panel

    override fun createActions(): Array<javax.swing.Action> = emptyArray()

    private fun setupActions() {
        btnGenerate.addActionListener {
            val className = tfClassName.text
            val inputText = jTextAreaInput.text
            if (className.isNullOrBlank()) {
                Messages.showMessageDialog("Class name cannot be empty.", "Error", Messages.getInformationIcon())
                return@addActionListener
            }
            if (inputText.isNullOrBlank()) {
                Messages.showMessageDialog("Input text cannot be empty.", "Error", Messages.getInformationIcon())
                return@addActionListener
            }
            val validationError = getInputValidationError()
            if (validationError != null) {
                Messages.showMessageDialog(validationError, "Error", Messages.getInformationIcon())
                return@addActionListener
            }
            onClickListener?.onGenerate(fieldParseConfig, className.trim(), inputText)
            close(OK_EXIT_CODE)
        }
        btnCancel.addActionListener {
            onClickListener?.onCancel()
            close(CANCEL_EXIT_CODE)
        }
        btnSetting.addActionListener {
            try {
                val settingsDialog = GenerateFieldSettingsDialog()
                settingsDialog.show()
                fieldParseConfig = ConfigComponent.getInstance().resolveFieldParseConfig()
                updateGenerateButtonState()
            } catch (ex: Exception) {
                LOG.error("Open settings dialog failed", ex)
                Messages.showErrorDialog(contentPanel, ex.message ?: "Failed to open settings.", "Settings")
            }
        }
    }

    private fun initUiComponents(initialClassName: String?): JPanel {
        val container = JPanel(BorderLayout(0, JBUI.scale(UI.Spacing.DIALOG_CONTENT_GAP)))
        container.preferredSize = DIALOG_SIZE
        container.border = JBUI.Borders.empty(
            UI.Spacing.DIALOG_TOP_PADDING,
            UI.Spacing.DIALOG_HORIZONTAL_PADDING,
            UI.Spacing.DIALOG_BOTTOM_PADDING,
            UI.Spacing.DIALOG_HORIZONTAL_PADDING
        )

        container.add(createHeaderPanel(), BorderLayout.NORTH)

        jTextAreaInput = JTextArea().apply {
            font = font.deriveFont(font.size2D + UI.FontSize.INPUT_TEXT)
        }
        val scrollPane: JScrollPane = LineNumberTextArea.wrap(jTextAreaInput)
        scrollPane.preferredSize = JBUI.size(UI.Size.MAIN_INPUT_WIDTH, UI.Size.MAIN_INPUT_HEIGHT)

        val centerPanel = JPanel(BorderLayout(0, JBUI.scale(UI.Spacing.DIALOG_CONTENT_GAP)))
        centerPanel.add(scrollPane, BorderLayout.CENTER)
        centerPanel.add(createInputOptionsPanel(initialClassName), BorderLayout.SOUTH)
        container.add(centerPanel, BorderLayout.CENTER)

        container.add(createButtonPanel(), BorderLayout.SOUTH)
        bindValidation()
        updateGenerateButtonState()
        return container
    }

    private fun createInputOptionsPanel(initialClassName: String?): JPanel {
        val optionsPanel = JPanel()
        optionsPanel.isOpaque = false
        optionsPanel.layout = BoxLayout(optionsPanel, BoxLayout.Y_AXIS)
        optionsPanel.add(createInputValidationPanel())
        optionsPanel.add(Box.createVerticalStrut(JBUI.scale(UI.Spacing.DIALOG_CONTENT_GAP)))
        optionsPanel.add(createClassNamePanel(initialClassName))
        return optionsPanel
    }

    private fun createHeaderPanel(): JPanel {
        val headerPanel = JPanel(BorderLayout(JBUI.scale(UI.Spacing.ROW_GAP_COMPACT), 0))
        headerPanel.isOpaque = false

        val titleLabel = JLabel("Generate fields from text:")
        titleLabel.font = titleLabel.font.deriveFont(titleLabel.font.size2D + UI.FontSize.CONTENT_TEXT)
        headerPanel.add(titleLabel, BorderLayout.WEST)
        headerPanel.add(createRepositoryLinkLabel(), BorderLayout.EAST)
        return headerPanel
    }

    private fun createRepositoryLinkLabel(): JLabel {
        val linkLabel = JLabel("<html><a href=\"$REPOSITORY_URL\">$REPOSITORY_URL</a></html>")
        linkLabel.cursor = Cursor.getPredefinedCursor(Cursor.HAND_CURSOR)
        linkLabel.addMouseListener(object : MouseAdapter() {
            override fun mouseClicked(e: MouseEvent) {
                BrowserUtil.browse(REPOSITORY_URL)
            }
        })
        return linkLabel
    }

    private fun createInputValidationPanel(): JPanel {
        val validationPanel = JPanel()
        validationPanel.isOpaque = false
        validationPanel.layout = BoxLayout(validationPanel, BoxLayout.Y_AXIS)
        validationPanel.alignmentX = JComponent.LEFT_ALIGNMENT

        chkRegexValidation = JCheckBox("Validate input with regex")
        chkRegexValidation.isSelected = fieldParseConfig.isEnableInputRegexValidation
        chkRegexValidation.alignmentX = JComponent.LEFT_ALIGNMENT
        validationPanel.add(chkRegexValidation)
        return validationPanel
    }

    private fun createClassNamePanel(initialClassName: String?): JPanel {
        val classNamePanel = JPanel(BorderLayout(JBUI.scale(UI.Spacing.ROW_GAP_COMPACT), 0))
        classNamePanel.isOpaque = false
        classNamePanel.alignmentX = JComponent.LEFT_ALIGNMENT
        classNamePanel.preferredSize = JBUI.size(0, UI.Size.INPUT_HEIGHT)
        classNamePanel.minimumSize = JBUI.size(0, UI.Size.INPUT_HEIGHT)

        val label = JLabel("Class Name")
        label.horizontalAlignment = SwingConstants.LEFT
        label.font = label.font.deriveFont(label.font.size2D + UI.FontSize.CONTENT_TEXT)
        classNamePanel.add(label, BorderLayout.WEST)

        tfClassName = JTextField(initialClassName.orEmpty())
        tfClassName.font = tfClassName.font.deriveFont(tfClassName.font.size2D + UI.FontSize.INPUT_TEXT)
        tfClassName.margin = JBUI.insets(
            UI.Spacing.INPUT_VERTICAL_PADDING,
            UI.Spacing.INPUT_HORIZONTAL_PADDING,
            UI.Spacing.INPUT_VERTICAL_PADDING,
            UI.Spacing.INPUT_HORIZONTAL_PADDING
        )
        val inputSize = Dimension(Int.MAX_VALUE, JBUI.scale(UI.Size.INPUT_HEIGHT))
        tfClassName.preferredSize = inputSize
        tfClassName.minimumSize = inputSize
        tfClassName.maximumSize = inputSize
        classNamePanel.add(tfClassName, BorderLayout.CENTER)
        return classNamePanel
    }

    private fun createButtonPanel(): JPanel {
        val buttonPanel = JPanel(BorderLayout())
        buttonPanel.isOpaque = false
        btnSetting = JButton("Settings")
        buttonPanel.add(btnSetting, BorderLayout.WEST)

        val actionButtonPanel = JPanel()
        actionButtonPanel.isOpaque = false
        actionButtonPanel.layout = BoxLayout(actionButtonPanel, BoxLayout.X_AXIS)
        btnGenerate = JButton("Generate")
        btnCancel = JButton("Cancel")
        actionButtonPanel.add(btnGenerate)
        actionButtonPanel.add(Box.createHorizontalStrut(JBUI.scale(UI.Spacing.BUTTON_GAP)))
        actionButtonPanel.add(btnCancel)
        buttonPanel.add(actionButtonPanel, BorderLayout.EAST)
        return buttonPanel
    }

    private fun applyPrimaryButtonStyle() {
        btnGenerate.putClientProperty("JButton.buttonType", "default")
        rootPane.defaultButton = btnGenerate
    }

    private fun bindValidation() {
        val listener = object : DocumentListener {
            override fun insertUpdate(e: DocumentEvent) = updateGenerateButtonState()
            override fun removeUpdate(e: DocumentEvent) = updateGenerateButtonState()
            override fun changedUpdate(e: DocumentEvent) = updateGenerateButtonState()
        }
        jTextAreaInput.document.addDocumentListener(listener)
        tfClassName.document.addDocumentListener(listener)
        chkRegexValidation.addActionListener {
            fieldParseConfig.isEnableInputRegexValidation = chkRegexValidation.isSelected
            updateGenerateButtonState()
        }
    }

    private fun updateGenerateButtonState() {
        btnGenerate.isEnabled = !(
            tfClassName.text.isNullOrBlank() ||
                jTextAreaInput.text.isNullOrBlank() ||
                getInputValidationError() != null
            )

    }

    private fun getInputValidationError(): String? {
        if (!chkRegexValidation.isSelected) {
            return null
        }
        val inputText = jTextAreaInput.text
        for (line in inputText.split(Regex("\\r?\\n"))) {
            if (line.isBlank()) {
                continue
            }
            if (!INPUT_VALIDATION_PATTERN.matcher(line).matches()) {
                return "Input text does not match the validation regex: $INPUT_VALIDATION_REGEX"
            }
        }
        return null
    }

    private fun registerEscapeAction() {
        panel.registerKeyboardAction(
            {
                onClickListener?.onCancel()
                close(CANCEL_EXIT_CODE)
            },
            javax.swing.KeyStroke.getKeyStroke(java.awt.event.KeyEvent.VK_ESCAPE, 0),
            JComponent.WHEN_ANCESTOR_OF_FOCUSED_COMPONENT
        )
    }

    fun setOnClickListener(onClickListener: OnClickListener?) {
        this.onClickListener = onClickListener
    }

    interface OnClickListener {
        fun onGenerate(fieldParseConfig: FieldParseConfig, className: String, text: String)
        fun onCancel()
    }

    companion object {
        private val DIALOG_SIZE = JBUI.size(UI.Size.MAIN_DIALOG_WIDTH, UI.Size.MAIN_DIALOG_HEIGHT)
        private val LOG = Logger.getInstance(GenerateFieldDialog::class.java)
        private const val INPUT_VALIDATION_REGEX = "[^\\t]+\\t[^\\t]*(?:\\t[^\\t]*)*"
        private val INPUT_VALIDATION_PATTERN: Pattern = Pattern.compile(INPUT_VALIDATION_REGEX)
        private const val REPOSITORY_URL = "https://github.com/jenly1314/JvmFieldGenerator"
    }
}
